### ISSUE TITLE
xs:extension in xs:redefine leads to "wrong redefinition without self-reference"

### ISSUE THREAD

**Issue by leo-b** [sissaschool/xmlschema#265]
Hi!
I am trying to redefine a complex type using `xs:redefine` together with `xs:extension`:

```xml
  <xs:redefine schemaLocation="dvu-schema-intermediate.xsd">
    <xs:complexType name="FullStudentInfo">
      <xs:complexContent>
	<xs:extension base="UniStudentInfo">
          <xs:sequence>
	    <xs:element name="studienbeitrag" type="KontostandUniInfo" minOccurs="0"/>
	  </xs:sequence>
	</xs:extension>
      </xs:complexContent>
    </xs:complexType>
  </xs:redefine>
```

Unfortunately this fails with
```
  File "/usr/local/lib/python3.9/site-packages/xmlschema/validators/complex_types.py", line 230, in _parse
    self.parse_error("wrong redefinition without self-reference")
  File "/usr/local/lib/python3.9/site-packages/xmlschema/validators/xsdbase.py", line 183, in parse_error
    raise error
xmlschema.validators.exceptions.XMLSchemaParseError: wrong redefinition without self-reference:

Schema:

  <xs:complexType xmlns:xs="http://www.w3.org/2001/XMLSchema" name="FullStudentInfo">
    <xs:complexContent>
      <xs:extension base="UniStudentInfo">
        <xs:sequence>
          <xs:element name="studienbeitrag" type="KontostandUniInfo" minOccurs="0" />
        </xs:sequence>
      </xs:extension>
    </xs:complexContent>
  </xs:complexType>

Path: /xs:schema/xs:redefine/xs:complexType
```

The debugger shows:
```
(Pdb) self.elem
<Element '{http://www.w3.org/2001/XMLSchema}complexType' at 0x7f30ff30d9f0>
(Pdb) derivation_elem
<Element '{http://www.w3.org/2001/XMLSchema}extension' at 0x7f30ff30db30>
(Pdb) self.name     
'{http://www.brz.gv.at/datenverbund-unis}FullStudentInfo'
(Pdb) self.base_type.name
'{http://www.brz.gv.at/datenverbund-unis}UniStudentInfo'
(Pdb) self.redefine
XsdComplexType(name='FullStudentInfo')
```

I am not a XML Schema expert but I guess this should work since it is only the _base_ and not the redefined type.

Is this the intended behavior?

P.S.: I can provide the whole example of course, if this is useful.

**Comment by brunato** [comment]
Hi,
redefinitions need a reference to itself (FullStudentInfo), so you have to encapsulate your redefined extension into another one that reference to itself. 

XSD 1.1 has xs:overrides that does not have these limitations.

**Comment by leo-b** [comment]
Hi Davide!
Sorry to misuse this issue as a support request..

Unfortunately I don't understand how I could do the encapsulation correctly.

I'd appreciate if you could help with a practical example.

My problem is that my schema should derive from another one that I am not able to modify directly:

### original (read-only) schema

```xml
	<xs:complexType name="FullStudentInfo">
		<xs:complexContent>
			<xs:extension base="UniStudentInfo">
				<xs:sequence>
					<xs:element name="studienliste" type="studienliste"/>
					<xs:element name="studienbeitrag" type="KontostandUniInfo"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>
```

`UniStudentInfo` is a sequence that itself extends another sequence and so on.
I need to change the element `studienbeitrag` to `minOccurs="0"`.

My approach was to use an intermediate schema to redefine `FullStudentinfo` with an empty sequence and re-add the sequence elements in another schema. However I don't see how to rebuild the sequence elements without using `<xs:extension base="UniStudentInfo">`.

Do you have an idea how to solve this?
cheers,
--leo

**Comment by brunato** [comment]
If i don't remember wrong it should be:

```xsd
      <xs:complexType name="FullStudentInfo">
        <xs:complexContent>
          <xs:restriction base="FullStudentInfo">
            <xs:sequence>
              <xs:element name="studienbeitrag" type="KontostandUniInfo"/>
            </xs:sequence>
          </xs:restriction>
        </xs:complexContent>
      </xs:complexType>
```
eventually adding missing _UniStudentInfo_'s elements.

If it's not, post also the other involved types of your source schema, to let me verify the correct form.

**Comment by leo-b** [comment]
> If i don't remember wrong it should be

This will lead to a new sequence with only one element (`studienbeitrag`).

> eventually adding missing UniStudentInfo's elements.

Yes. I guess I need to add all elements of all derived types again. Unfortunately the types are nested, which makes it quite cumbersome.
The involved types are:

```xml
	<xs:complexType name="studentInfoHuelle" abstract="true"/>

	<xs:complexType name="studentInfoMinimal">
		<xs:complexContent>
			<xs:extension base="studentInfoHuelle">
				<xs:sequence>
					<xs:element name="vorname" type="sbtTextNotEmpty" minOccurs="1"/>
					<xs:element name="nachname" type="sbtTextNotEmpty" minOccurs="1"/>
					<xs:element name="geburtsdatum" type="xs:date" minOccurs="1"/>
					<xs:element name="geschlecht" type="geschlechtTyp" minOccurs="1"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>

	<xs:complexType name="studentInfoBasis">
		<xs:complexContent>
			<xs:extension base="studentInfoMinimal">
				<xs:sequence>
					<xs:element name="staatsbuergerschaft" type="staatTyp" minOccurs="0"/>
					<xs:element name="akadgrad" type="akadTyp" minOccurs="0"/>
					<xs:element name="akadnach" type="akadTyp" minOccurs="0"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>

	<xs:complexType name="studentInfoExtended">
		<xs:complexContent>
			<xs:extension base="studentInfoBasis">
				<xs:sequence>
					<xs:element name="bpk" type="bpkTyp" minOccurs="0"/>
					<xs:element name="svnr" type="svnrTyp" minOccurs="0"/>
					<xs:element name="ekz" type="ekzTyp" minOccurs="0"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>

	<xs:complexType name="studentInfoBasisAdressen">
		<xs:complexContent>
			<xs:extension base="studentInfoExtended">
				<xs:sequence>
					<xs:element name="adressen" type="adressenListe"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>

	<xs:complexType name="UniStudentInfo">
		<xs:complexContent>
			<xs:extension base="studentInfoBasisAdressen">
				<xs:sequence>
					<xs:element name="studierendenkey" type="Studierendenkey" minOccurs="1" maxOccurs="1"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>

```

Thanks for your help!
--leo

**Comment by brunato** [comment]
This would be the redefinition for your case:

```xsd
  <xs:redefine schemaLocation="issue_265-1.xsd">
      <xs:complexType name="FullStudentInfo">
        <xs:complexContent>
          <xs:restriction base="FullStudentInfo">
            <xs:sequence>
              <xs:element name="vorname" type="sbtTextNotEmpty" minOccurs="1"/>
              <xs:element name="nachname" type="sbtTextNotEmpty" minOccurs="1"/>
              <xs:element name="geburtsdatum" type="xs:date" minOccurs="1"/>
              <xs:element name="geschlecht" type="geschlechtTyp" minOccurs="1"/>
              <xs:element name="staatsbuergerschaft" type="staatTyp" minOccurs="0"/>
              <xs:element name="akadgrad" type="akadTyp" minOccurs="0"/>
              <xs:element name="akadnach" type="akadTyp" minOccurs="0"/>
              <xs:element name="bpk" type="bpkTyp" minOccurs="0"/>
              <xs:element name="svnr" type="svnrTyp" minOccurs="0"/>
              <xs:element name="ekz" type="ekzTyp" minOccurs="0"/>
              <xs:element name="adressen" type="adressenListe"/>
              <xs:element name="studierendenkey" type="Studierendenkey" minOccurs="1" maxOccurs="1"/>
              <xs:element name="studienbeitrag" type="KontostandUniInfo"/>
            </xs:sequence>
          </xs:restriction>
        </xs:complexContent>
      </xs:complexType>
  </xs:redefine>
```

But (un)fortunately i found a bug on complexType redefinition and this cannot be a legal restriction ...

To be legal the removed element should be emptiable in the base model, so like this:

```xsd
      <xs:complexType name="FullStudentInfo">
		<xs:complexContent>
			<xs:extension base="UniStudentInfo">
				<xs:sequence>
					<xs:element name="studienliste" type="studienliste" minOccurs="0"/>
					<xs:element name="studienbeitrag" type="KontostandUniInfo"/>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>
```

Otherwise with `xmlschema.XMLSchema11` you can use an `xs:override`, that could be like your initial version:

```xsd
  <xs:override schemaLocation="issue_265-1.xsd">
    <xs:complexType name="FullStudentInfo">
      <xs:complexContent>
	<xs:extension base="UniStudentInfo">
          <xs:sequence>
	    <xs:element name="studienbeitrag" type="KontostandUniInfo" minOccurs="0"/>
	  </xs:sequence>
	</xs:extension>
      </xs:complexContent>
    </xs:complexType>
  </xs:override>
```

**Comment by leo-b** [comment]
> Otherwise with xmlschema.XMLSchema11 you can use an xs:override, that could be like your initial version

Oh - wow! 
I didn't know that it is possible to mix schema versions, i.e. to override 1.0 definitions in a 1.1 schema!

I already came to the conclusion that `xs:override` would be the simplest and cleanest way to solve this but I abandoned this path because the original (read-only) schema is 1.0.

Thanks a lot for your help!
It works like a charm!

Cheers,
--leo
